packages/core/workflows-sdk/src/helper/workflow-export.ts

```diff
@@ -448,3 +461,43 @@ export const exportWorkflow = <TData = unknown, TResult = unknown>(
   MedusaWorkflow.registerWorkflow(workflowId, exportedWorkflow)
   return exportedWorkflow as MainExportedWorkflow<TData, TResult>
 }
+
+function attachOnFinishReleaseEvents(
+  events: DistributedTransactionEvents = {},
+  eventGroupId: string,
+  flow: LocalWorkflow
+) {
+  const onFinish = events.onFinish
+
+  const wrappedOnFinish = async (args: {
+    transaction: DistributedTransaction
+    result?: unknown
+    errors?: unknown[]
+  }) => {
+    await onFinish?.(args)
+
+    const eventBusService = (flow.container as MedusaContainer).resolve(
+      ModuleRegistrationName.EVENT_BUS,
+      { allowUnregistered: true }
+    )
+
+    if (!eventBusService || !eventGroupId) {
+      return
+    }
+
+    const { transaction } = args
+    const failedStatus = [TransactionState.FAILED, TransactionState.REVERTED]
+
+    if (failedStatus.includes(transaction.getState())) {
+      return await eventBusService.clearGroupedEvents(eventGroupId)
+    }
+
+    await eventBusService
+      .releaseGroupedEvents(eventGroupId)
+      .catch(async (e) => {
+        await flow.cancel(transaction)
```

Comment:
question: Do we need to log or something?

Reply:
Sorry @olivermrbl but again we are probably doing things at the same time, me adding tests and improvements and you guys commenting 😂 so it is already in